Add Modbus cover

vzahradnik opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
<!--## Breaking change-->
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
<!--
In order to add support for the new `Cover` integration, we had to change the configuration format for the existing `Modbus` platform. `Cover` entries will be defined under new `Modbus` subsection. Because YAML expected a list, it wasn't possible to add a new subsection without breaking the YAML format. Therefore, all `hubs` will be defined under the new `hubs` subsection.

Before:
```yaml
modbus:
  - name: plc_01
    type: tcp
    host: 192.168.1.31
    port: 502
```

After:
```yaml
modbus:
  hubs:
    - name: plc_01
      type: tcp
      host: 192.168.1.31
      port: 502

  covers:
    ...
```
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This pull request adds a new Modbus entity - cover. At the moment, this implementation supports `open`, `closed`, `opening`, and `closing` states.

User can control the cover using a coil or a holding register. If a coil is used, we can't detect intermediary states like *opening* or *closing*. Therefore, an optional `status_register` attribute was defined, so that the cover state can be read from the register.

If user decides to use holding register to control the cover, this register is also used for reading the states back. However, user can specify `status_register` which will take priority also in this case.

To map register values to actual states, four optional attributes were defined (with reasonable default values). User can override any or all of the values, e.g., `state_open=1`, or `state_closing=3`. If the cover is controlled by the holding register, the `state_open` and `state_closed` attributes are also used for writing into the register, and controlling the state of the cover.

Configuration validation was adjusted, so that either `coil` or `register` is specified. User needs to specify one of these attributes. If both or none of them are specified, a config error is logged.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
modbus:
    - name: hub1
      type: tcp
      host: IP_ADDRESS
      port: 502

      covers:
        - name: Door1
          hub: hub1
          slave: 1
          coil: 1
          scan_interval: 10
          device_class: door
        - name: Door2
          hub: hub1
          slave: 2
          coil: 2
          device_class: door
```
Note: for more examples, please refer to the [documentation PR](home-assistant/home-assistant.io#12687). I provided several examples to make the usage clear to the user.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#12687

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

vzahradnik committed:
Add Modbus cover

janiversen reviewed (approved):
We should think of moving the read functions into modbushub in __init__.py as it seems there are most (if not all) entities, use the exact same pattern (inclduding the testing for HOLDING or INPUT). However this can be done in another PR.

The code looks good, good work!

janiversen commented:
Your 3 PR are all OK from my side, now we just need to wait for someone with write access to look at them.

Regarding the score, once I am over the current set of errors (open issues), I will deal with the logging.

vzahradnik commented:
@janiversen what's the status with the async `pymodbus` code? I'm thinking of making this code synchronous again just for the sake of better maintainability. It shouldn't be much work, and at least we could use my components.

janiversen commented:
Sorry for the delay in responding, I was occupied with the change in “state of alert” here in Spain, which are slowly moving back to a somewhat normal life.

The async code is waiting for a new version from pymodbus, I am working on an alternative that bypasses the communication part of pymodbus, but it is only moving slowly forward.

I prefer if it is not much work, that you convert to sync and we get it merged. When in future we make an async version I will make the needed changes for all components.

vzahradnik commented:
@janiversen I was busy with other stuff too. But I will start working on it again. Converting my code shouldn't be much work. Also, I'll keep async versions. Later, either you could modify it or just ask me and I'll provide the files.

janiversen commented:
Perfect, let me know when you want a review, so we can get this code merged, it is long overdue.

janiversen reviewed (approved):
Ready to be merged.

vzahradnik commented:
@janiversen I still work on this. I didn't verify my latest changes manually. Also, I would love to include at least some unit tests, but I didn't grasp how existing tests are written. Tests for Modbus sensor are clear, but cover is more complicated. My main issue is that the tests in general are not well documented, and sometimes I have no clue what they're doing.

However, this PR can be changed from draft again, and you guys can start reviewing my code. I can always include tests in a separate PR.

Thanks!

bradkeifer commented:
> Sorry for the delay in responding, I was occupied with the change in “state of alert” here in Spain, which are slowly moving back to a somewhat normal life.
> 
> The async code is waiting for a new version from pymodbus, I am working on an alternative that bypasses the communication part of pymodbus, but it is only moving slowly forward.
> 
> I prefer if it is not much work, that you convert to sync and we get it merged. When in future we make an async version I will make the needed changes for all components.

@janiversen - If I can help with testing of the async solution, please let me know.

vzahradnik commented:
@janiversen I fixed a couple of issues, and tested the code manually. Now it's ready for review. Please take a look at it now. Thanks!

janiversen commented:
Looks real good, ready to be merged from my side.

Lets see if @MartinHjelmare is happy too.

vzahradnik committed:
Fix improper commands written for Modbus cover via coil

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,252 @@
+):
+    """Read configuration and create Modbus cover."""
+    for cover in config[CONF_COVERS]:
+        async_add_entities([ModbusCover(hass, cover)])
```
Add the entities to a list first and then make single call to `async_add_entities`.

vzahradnik replied:
Done.

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,252 @@
+    """Representation of a Modbus cover."""
+
+    def __init__(
+        self, hass: HomeAssistantType, config: Dict[str, Any],
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant. Pass in the hub instead.

vzahradnik replied:
Done, thanks for pointing that out.

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,252 @@
+                        vol.Exclusive(CONF_REGISTER, CONF_INPUT_TYPE): cv.positive_int,
+                    }
+                ),
+                COVER_SCHEMA.extend(
```
Why extend the same schema twice?

vzahradnik replied:
I didn't find a better way back then. Thanks for the `cv.has_at_least_one_key`, it made validation much more simple now.

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,252 @@
+                COVER_SCHEMA.extend(
+                    {
+                        vol.Required(
+                            vol.Any(CALL_TYPE_COIL, CONF_REGISTER)
```
Use `cv.has_at_least_one_key` instead.

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,252 @@
+            self._available = False
+            return
+
+        if isinstance(result, (ModbusException, ExceptionResponse)):
```
Side note: It's weird that the exception is returned instead of raised.

vzahradnik replied:
Yes, it is. @balloob said the same some time ago. Once we migrate to async code, we'll get rid of this line.

MartinHjelmare commented on tests/components/modbus/test_cover.py:
```diff
@@ -0,0 +1 @@
+"""Tests for the Modbus cover component."""
```
Tests are yet missing.

vzahradnik replied:
I removed the empty test file.

MartinHjelmare reviewed (changes requested):
Looks good! Some comments.

vzahradnik committed:
Make changes per review comments

vzahradnik commented:
@MartinHjelmare I implemented changes according to your comments. Please check now. Thanks!

MartinHjelmare reviewed (approved):
Nice!

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,250 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Sorry, I didn't think about this: We want the config to be hosted in the main integration for new platforms. What's the plan for that?

See https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md

vzahradnik replied:
I didn't know about that, but I can rework the code. Could you provide some examples of platforms, which implement this new approach?

So, if I understand, the only difference is to move the config from `cover` section directly under modbus section, and to initialize covers from there. Something like:

```yaml
modbus:
  - name: plc_01
    type: tcp
    host: 127.0.0.1
    port: 5020
    covers:
        - name: plc_01_door
          hub: plc_01
          slave: 1
          device_class: gate
          coil: 0
          state_opening: 1
          state_open: 2
          state_closing: 3
          state_closed: 4
```

MartinHjelmare replied:
Yeah, and then we need to load the cover platform via discovery with `load_platform`:

https://github.com/home-assistant/core/blob/f6df85f8cef4ac1b53eb66235f4f6a0c5acec513/homeassistant/helpers/discovery.py#L130-L137

The config can be passed either via `discovery_info` parameter or via `hass.data` to the platform.

vzahradnik replied:
Thanks, I found a integration with this approach. Now I'm adapting the code. Documentation PR needs to be updated as well. I'll post an update here.

vzahradnik committed:
Fix default hub not defined

Since support for multiple hubs was added, the default hub option
was not implemented correctly. Now I added necessary logic to make
it work. First hub in a list will be used as a default hub.

vzahradnik commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -95,6 +94,10 @@ def setup(hass, config):
     for client_config in config[DOMAIN]:
         hub_collect[client_config[CONF_NAME]] = ModbusHub(client_config)
 
+        # Set first hub in a list as a default
```
@janiversen please check this code. I just found a bug, which was in the Modbus code for a very long time. Each entity has an optional parameter `hub`. If it's not specified, a `DEFAULT_HUB` value gets assigned instead.

When such an entity wants to access a "default" hub, it looks into Modbus hub dictionary and asks for a "default" hub, which doesn't exist. This small change fixes that as I'm assigning the first defined hub in a list as a default, i.e., two keys point to one `ModbusHub` instance.

In case you'll think of a better way how to handle this situation, please let me know. One option would be to get rid of the `default hub` completely, and make it a required parameter.

janiversen replied:
My preference is to drop “default hub” and make hub required. this is not a real breaking change since default hub dis not work (good work catching that bug<).

vzahradnik replied:
Regarding default hub: the code actually works if user omits hub name in the hub definition. Anyway, I prepared a separate PR to discuss this issue there. In my opinion, it's still better to make this parameter mandatory. However, it is a **breaking change**.

In cover I will enforce the `hub` parameter as required.

Reference: home-assistant/core#37546

vzahradnik committed:
Move Cover config under Modbus section

vzahradnik commented:
@MartinHjelmare, @janiversen please take a look at the latest code.
- Reworked to read cover config from the `Modbus` section
- Fixed a bug with non-existing `DEFAULT_HUB`

## CONFIG CHANGE
Due to the change to move `cover` config inside `Modbus` section, I had to introduce a **breaking change**. Originally, under `Modbus` we list all the hubs. Because I needed to add a `cover` section, the list of hubs was moved under a new config entry called `hubs`. If the `cover` section followed the list of hubs directly, such a YAML file was invalid.

For a better demonstration, here's the original config:
```yaml
modbus:
  - name: plc_01
    type: tcp
    host: 127.0.0.1
    port: 5020
```

And here's a new config. Note the `hubs` and `covers` sections:
```yaml
modbus:
  hubs:
    - name: plc_01
      type: tcp
      host: 127.0.0.1
      port: 5020

  covers:
    - name: plc_01_door
      scan_interval: 1
      hub: plc_01
      slave: 1
      coil: 0
      device_class: door
      status_register: 4
      status_register_type: input
      state_opening: 1
      state_open: 2
      state_closing: 3
      state_closed: 4
```

This config is very readable, but it's a breaking change. What's your suggestion on this? Perhaps I could introduce a backward compatibility. However, in such cases users cannot use the new `cover`. Such YAML file is invalid (checked with a YAML validator).

### Warnings
When I run such a config, Home Assistant issues a warning. It's not critical and has no effect on the component. However, please take a look at the log output.

```bash
2020-07-03 17:40:51 WARNING (Recorder) [homeassistant.components.recorder] Event is not JSON serializable: <Event platform_discovered[L]: service=load_platform.cover, platform=modbus, discovered=[OrderedDict([('name', 'plc_01_door'), ('scan_interval', datetime.timedelta(seconds=1)), ('hub', 'plc_01'), ('slave', 1), ('coil', 0), ('device_class', 'door'), ('status_register', 4), ('status_register_type', 'input'), ('state_opening', 1), ('state_open', 2), ('state_closing', 3), ('state_closed', 4)])]>
```

## DEFAULT HUB FIX
Each modbus component has an optional `hub` parameter. If not specified, it uses a `DEFAULT_HUB`. However, this hub is never initialized and such code fails. I introduced a small fix, where I assign first hub from the config to be default. However, because nobody reported an error (as far as I know), I assume all users define this parameter. I think it would be better to make this parameter required to eliminate ambiguity. This would be a breaking change, too.

Please advise on both issues.

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
The scan interval is not json serializable. Either convert it to eg seconds before passing it in the discovery info or store the config in hass.data instead.

MartinHjelmare commented:
Please add a breaking change paragraph to the PR description describing what changed and what the user needs to do to cope with the breaking change.

janiversen commented:
> The scan interval is not json serializable. Either convert it to eg seconds before passing it in the discovery info or store the config in hass.data instead.

please make it seconds, which is used in all other configs.

Code looks good, lets get it merged (without default hub<)

janiversen commented:
RemRK the CI faIlure seems real:

homeassistant/components/modbus/cover.py:40:8: R1704: Redefining argument with the local name 'config' (redefined-argument-from-local)

please change variable name.

janiversen commented:
Sorry just saw you requested comments on 2 changes.

It would be nice (but NOT a demand) to provide backward compatibility with a big warning.

vzahradnik committed:
Revert setting up a default hub alias

vzahradnik committed:
Make hub mandatory for Cover

vzahradnik committed:
Add default scan interval

vzahradnik committed:
Read scan_interval from discovery info

vzahradnik committed:
Fix linter error

vzahradnik commented on homeassistant/helpers/entity_component.py:
```diff
@@ -229,9 +229,23 @@ async def async_setup_platform(
-        scan_interval = platform_config.get(
-            CONF_SCAN_INTERVAL, getattr(platform, "SCAN_INTERVAL", None)
-        )
+        # FIXME: Components don't pass discovery_info in the correct format:
```
@MartinHjelmare, @janiversen When we setup a platform through dynamic loading, current code ignores `scan_interval` config from `discovery_info`. Following code fixes that and now it works as expected.

Caveats:
- Platform code expects time interval to be defined as `datetime.timedelta(seconds=...)`. I can rework my code to use integer in my structure, however, in such a case we should do a conversion to timedelta in this place. Validation method `cv.time_period` does the conversion for us. The downside is `Recorder` warning, as mentioned.
```bash
2020-07-06 14:09:14 WARNING (Recorder) [homeassistant.components.recorder] Event is not JSON serializable: <Event platform_discovered[L]: service=load_platform.cover, platform=modbus, discovered=[OrderedDict([('name', 'plc_01_door'), ('hub', 'plc_01'), ('slave', 1), ('device_class', 'door'), ('scan_interval', datetime.timedelta(seconds=5)), ('coil', 0), ('status_register', 4), ('status_register_type', 'input'), ('state_opening', 1), ('state_open', 2), ('state_closing', 3), ('state_closed', 4)])]>
```
- When components set up platform, they pass `discovery_info` in different formats. Most components pass `None` when not initialized dynamically, however, the `Updater` component passes `{}`. Also, `cover` component loaded dynamically passes a list of a single dictionary, instead of passing the dictionary directly (expected `Optional[DiscoveryInfoType]`, which is `Dict[str, Any]`)
- I updated the code to make it work. However, the format of `DiscoveryInfoType` should be unified across the whole codebase. Temporarily, I changed the type to `Optional[Any]` to supress linter complaint.

Here's my debugging log. First line is the name of the component, second line is the content of `discovery_info`.
```bash
[vzahradnik@Ryzen home-assistant-core (feature/modbus-cover)]$ hass
homeassistant
None
updater
{}
modbus
None
modbus
None
modbus
None
modbus
[OrderedDict([('name', 'plc_01_door'), ('hub', 'plc_01'), ('slave', 1), ('device_class', 'door'), ('scan_interval', datetime.timedelta(seconds=5)), ('coil', 0), ('status_register', 4), ('status_register_type', 'input'), ('state_opening', 1), ('state_open', 2), ('state_closing', 3), ('state_closed', 4)])]
```

vzahradnik commented:
## Summary Update:
- I created a separate PR for the issue with default hub. home-assistant/core#37546
- Documentation updated
- Breaking change info added
- CI failures fixed

## Scan Interval
Please see my comment in the `entity_component.py` file.

TL;DR; Home Assistant was ignoring the `scan_interval` parameter if passed through `discovery_info`. Also, it expects the value to be in `timedelta` format. So I kept the format for the validation unchanged. It ensures proper validation and conversion to `timedelta`. The drawback is a warning with JSON serialization.

@janiversen, @MartinHjelmare please take a look at the changes and advise how to proceed. Currently, the code is working OK.

vzahradnik committed:
Use default scan interval from Cover platform

MartinHjelmare commented on homeassistant/helpers/entity_component.py:
```diff
@@ -229,9 +229,23 @@ async def async_setup_platform(
+        # - Cover passes a list containing a single dictionary when initialized
+        # etc.mm@PentaphyllumYubiKey
+        if isinstance(discovery_info, list):
+            scan_interval = discovery_info[0].get(
```
Integrations are responsible for the discovery info. It's not standardized and we don't support setting the scan interval via the discovery info.

Please remove these changes.

vzahradnik replied:
OK, but how I suppose to make a 1-to-1 transition into a new config? Scan interval is set under the platform like this:
```yaml
cover:
 - platform: modbus
   scan_interval: 10
```

This attribute is essential for my component, and as you see it's handled outside of my code. In the `cover` platform itself.

MartinHjelmare replied:
Why is the option essential?

The common scan interval option is legacy. If an integration needs to adjust the scan interval dynamically or heterogenously it has to take control of polling either via our coordinator helper or via a custom update logic using our event helper.

vzahradnik replied:
By default, `cover` polls for updates once in 15 seconds. In my case, I need to read data once in a second.
If the `scan_interval` is legacy, maybe it should be mentioned in the documentation.

Anyway, if there's another way how to control this, I can rework the code.

So basically, I will set my own attribute `scan_interval` and use your helper methods to pass it on? Could you point me to some examples? For starters, please let me know which helpers did you mean, so that I can look at the code.

Thanks!

vzahradnik replied:
I found some other integrations with the new approach. I'll try to modify my code and will notify you...

MartinHjelmare replied:
The lowest allowed scan interval is 5 seconds.

If all platform instances can have the same scan interval we can just define the `SCAN_INTERVAL` constant in the platform module and it will override the default scan interval.

vzahradnik committed:
Handle polling for Modbus cover directly inside entity

vzahradnik commented:
I reworked `cover` to handle polling by itself:
- Default scan interval is set to 15 seconds
- SCAN_INTERVAL entry scheme updated to `positive_int` to avoid serialization warnings. Later it's converted to `datetime.timedelta`
- Inside `ModbusCover` I added the `should_poll()` property, which returns False. Now, polling is under full control of the `cover`.

Polling is implemented by the following code:
```python
async_track_time_interval(
    self.hass, lambda arg: self.update(), self._scan_interval
)
```
I use lambda function, because the `async_track_time_interval` method passes an argument, which the `update()` method doesn't expect.

To write entity changes to the state machine, I call `self.async_write_ha_state()` as the last statement in the `update()` method.

I tested the following code on my setup, I have no errors or warnings.

@janiversen, @MartinHjelmare please take a look.

Thanks!

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -87,13 +104,52 @@
+            vol.Required(CONF_NAME): cv.string,
+            vol.Optional(
+                CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+            ): cv.positive_int,
```
This should be `cv.time_period` as primary validator. Convert to seconds with a second validator.

```py
vol.All(cv.time_period, lambda value: value.total_seconds())
```

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,238 @@
+        else:
+            self._value = self._read_status_register()
+
+        self.async_write_ha_state()
```
We need to use the sync api in sync context. Use `self.schedule_update_ha_state`.

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,238 @@
+        if self._coil is not None:
+            self._write_coil(True)
+        else:
+            self._write_register(self._state_open)
```
When `should_poll` is false, we need to schedule the state update specifically in the actuator methods if there's no feedback from the device.

vzahradnik replied:
I'm not sure what exactly should I do. At the moment, I schedule timer during initialization by calling `async_track_time_interval()`. Is there anything else I should schedule? What do you mean by the *actuator* methods?

MartinHjelmare replied:
`open_cover` is an actuator method, `close_over` etc as well. These will ask the device to do something, ie actuate.

After the device has done something we want home assistant to show the new state of the device in the entity representing that device feature. So we need to update the entity state with new info.

When default polling is on, the `update` method is called automatically by home assistant and then the entity is state is updated. When default polling is off, this doesn't happen and we need to update entity state manually as needed.

To update the entity state from sync context we can call `self.schedule_update_ha_state` from inside the entity.

vzahradnik replied:
I added the call to `self._update()` into the `open_cover()` and `close_cover()` methods. These methods only initiate some action on Cover, but the result is not immediate. The `self._update()` method fetches the current state, and on last line it stores the state by calling `self.schedule_update_ha_state()`

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -87,13 +104,52 @@
+                vol.Required(CONF_HUBS): vol.All(
+                    cv.ensure_list, [vol.Any(SERIAL_SCHEMA, ETHERNET_SCHEMA)]
+                ),
+                vol.Optional(CONF_COVERS): vol.All(cv.ensure_list, [COVERS_SCHEMA]),
```
Maybe it's better to put the cover schema inside the base hub schema. Then we won't make a breaking change and the covers will be coupled naturally to a hub.

vzahradnik replied:
I thought about this variant, too, and I think it's better this way. Covers list is defined only once, and each entry points to proper Modbus hub. To me it looks concise. Also, I have changes planned for `light` and `fan` entities (new integrations). Having all lights, covers and fans defined in one place is in my opinion much more readable.

Consider this:
```yaml
modbus:
  - name: plc_01
    type: tcp
    host: 192.168.1.31
    port: 502
    covers:
        - name: plc_01_door
          hub: plc_01
          ....
    lights:
        - name: plc_01_light
           hub: plc_01
          ...
    fans:
        - name: plc_01_fan
           hub: plc_01
          ...
    name: plc_02
    type: tcp
    host: 192.168.1.32
    port: 502
    covers:
        - name: plc_02_door
          hub: plc_02
          ....
    lights:
        - name: plc_02_light
           hub: plc_02
          ...
    fans:
        - name: plc_02_fan
           hub: plc_02
          ...
````

Versus this:
```yaml
modbus:
    hubs:
      - name: plc_01
        type: tcp
        host: 192.168.1.31
        port: 502
      - name: plc_02
        type: tcp
        host: 192.168.1.32
        port: 502

    covers:
        - name: plc_01_door
          hub: plc_01
          ....
        - name: plc_02_door
          hub: plc_02
          ....

    lights:
        - name: plc_01_light
           hub: plc_01
          ...
        - name: plc_02_light
           hub: plc_02
          ...

    fans:
        - name: plc_01_fan
           hub: plc_01
          ...
        - name: plc_02_fan
           hub: plc_02
          ...
```

Second config is much better readable, especially if there area more than two hubs.

MartinHjelmare replied:
We don't need to define hub under the covers etc in the first variant. I think that's better and we're avoiding a breaking change. The latter is very important.

vzahradnik replied:
I can always write a compatibility code, if this is the only problem.

- If people don't use the new `cover` component, their config will still work. I'll just print a warning.
- If they decide to use the new `cover` component, they need to rework their config anyway, and instructions already show what needs to be changed

MartinHjelmare replied:
We'll break all existing user configs since we have changed the config schema. We shouldn't do that if we don't need to. Please use the suggestion. That's also in line with other integrations.

janiversen commented:
I think you wanted this merged before the others, so let me repeat my question here, do you have any idea when you will have time to iron out the last details (CI) ?

I am currently updating the tests, because I really want tests for all types.

vzahradnik commented:
@janiversen I planned to work on it this week, but I was still busy on my paid project. I hope to have some more time next week.

I can start with the Climate, so that you can continue on fixing the bug.

I will let you know once I have something new.

Thanks!

vzahradnik committed:
Move covers under hub config

vzahradnik committed:
Fix for review comment

vzahradnik commented:
Finally, I updated this PR with some of the requested changes:
- Covers config was moved under hub configuration. Documentation PR was also updated
- Fixed one issue mentioned in the PR review

Please take a look at it, so we can finally start cleaning this PR.

Thanks!

janiversen commented:
LGTM.

janiversen commented:
I would really not like to have/make a global change to the config of the whole modbus, for a couple of reasons:
- I am keeping the async version updated in a seperate branch, and it seems pymodbus is moving as we need. So bigger changes are just a lot of work.
- The integrations that uses modbus will probably be affected by a global config change, so we might break more than just modbus. My preference would be to keep the old configuration way alongside the new one for a couple of point releases, making a LARGE warning when the old configuration is used.

vzahradnik commented:
> I would really not like to have/make a global change to the config of the whole modbus

The code as is implemented now doesn't break anything. It adds Modbus covers as an optional list inside hub definition (see updated description). Existing config should *NOT* be affected.

Global config change makes sense, but it can be postponed. However, we should do it at some point, because maintenance and coding new features is harder.

janiversen commented:
Yes a global change surely makes sense, but it must be timed !!

I saw your config did not break the existing ones, sorry for not mentioning that. 

It would be real nice to get PR merged.

MartinHjelmare commented on homeassistant/components/modbus/cover.py:
```diff
@@ -0,0 +1,240 @@
+        else:
+            self._write_register(self._state_closed)
+
+    def update(self):
```
It's probably good to rename this method so it's not mixed up with the default `update` method that is used for polling entities.

vzahradnik replied:
I renamed it to `def _update(self)`. This method is used only internally now anyway.

janiversen commented:
Cover is new, so how can this be a breaking change?

MartinHjelmare commented:
That's stale.

vzahradnik committed:
Call update() from Cover actuator methods

MartinHjelmare reviewed (approved):
Looks good!

vzahradnik commented:
@MartinHjelmare thanks! What's the next step? Should I squash the commits before we merge it?

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -6,39 +6,84 @@
+            vol.Required(CONF_NAME): cv.string,
+            vol.Optional(
+                CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+            ): cv.positive_int,
```
This should be cv.time_period as primary validator. Convert to seconds with a second validator.
```suggestion
            ): vol.All(cv.time_period, lambda value: value.total_seconds()),
```

vzahradnik replied:
I applied your suggestion, and tested the change. Everything seems to be working fine.

MartinHjelmare reviewed (changes requested):
Sorry, this old comment hadn't been addressed.

MartinHjelmare commented:
No, we'll squash when we merge.

vzahradnik committed:
Fix time validation

vzahradnik commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -6,39 +6,84 @@
+    vol.Schema(
+        {
+            vol.Required(CONF_NAME): cv.string,
+            vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): vol.All(
```
@MartinHjelmare I applied your suggestion, and tested the change. Everything seems to be working fine.

MartinHjelmare reviewed (approved):
Thanks!

makuser commented:
@vzahradnik may I ask which modbus devices this integration works with on your end? I might be looking for a couple hundred of these.

vzahradnik commented:
@makuser I use this integration with Controllino Mega, which is an industry-grade Arduino Mega. I have most of the logic there, and Home Assistant provides a frontend and automation logic to it. I also have some other Modbus devices, but even in this case the data goes through Controllino. It takes Modbus over TCP, and sends it as Modbus RTU over RS485 (serial).

This solution works very well. I am thinking of writing a blog post about it. In the meantime, however, feel free to ask.
